File: 2015/07/rs/src/main.rs
use regex::Regex;
use std::clone::Clone;
use std::collections::HashMap;

type Connections = HashMap<String, Connection>;

pub enum Operand {
    Scalar(u32),
    Wire(String),
}

impl Operand {
    fn new(value: String) -> Operand {
        if let Ok(scalar) = value.parse() {
            return Operand::Scalar(scalar);
        }
        Operand::Wire(value)
    }
}

impl Clone for Operand {
    fn clone(&self) -> Operand {
        match self {
            Operand::Scalar(v) => Operand::Scalar(*v),
            Operand::Wire(w) => Operand::Wire(String::from(w)),
        }
    }
}

pub enum Operation {
    And,
    Or,
    Lshift,
    Rshift,
}

impl Operation {
    fn new(value: &str) -> Operation {
        match value {
            "AND" => Operation::And,
            "OR" => Operation::Or,
            "LSHIFT" => Operation::Lshift,
            "RSHIFT" => Operation::Rshift,
            _ => panic!("Unrecognized operation {}", value),
        }
    }
}

pub enum Connection {
    Input(Operand),
    Not(Operand),
    Binary(Operand, Operand, Operation),
}

pub struct CircuitSolver {
    solutions: Box<HashMap<String, u32>>,
}

impl CircuitSolver {
    pub fn new(solutions: HashMap<String, u32>) -> CircuitSolver {
        CircuitSolver {
            solutions: Box::new(solutions),
        }
    }

    fn get_value_from_operand(&mut self, operand: &Operand, connections: &Connections) -> u32 {
        match operand {
            Operand::Scalar(value) => *value,
            Operand::Wire(wire) => {
                self.get_value_from_connection(wire, connections.get(wire).unwrap(), connections)
            }
        }
    }

    fn get_value_from_binary_connection(
        &mut self,
        operand1: &Operand,
        operand2: &Operand,
        operation: &Operation,
        connections: &Connections,
    ) -> u32 {
        let x = self.get_value_from_operand(&operand1, connections);
        let y = self.get_value_from_operand(&operand2, connections);
        match operation {
            Operation::And => x & y,
            Operation::Or => x | y,
            Operation::Lshift => x << y,
            Operation::Rshift => x >> y,
        }
    }

    fn calculate_value_for_connection(
        &mut self,
        connection: &Connection,
        connections: &Connections,
    ) -> u32 {
        match connection {
            Connection::Input(operand) => self.get_value_from_operand(operand, connections),
            Connection::Not(operand) => !self.get_value_from_operand(operand, connections),
            Connection::Binary(operand1, operand2, operation) => {
                self.get_value_from_binary_connection(operand1, operand2, operation, connections)
            }
        }
    }

    fn get_value_from_connection(
        &mut self,
        target: &str,
        connection: &Connection,
        connections: &Connections,
    ) -> u32 {
        if let Some(value) = self.solutions.get(target) {
            return *value;
        } else {
            let result = self.calculate_value_for_connection(connection, connections);
            self.solutions
                .entry(String::from(target.clone()))
                .or_insert(result);
            result
        }
    }

    pub fn solve_for(&mut self, target: &str, connections: &Connections) -> u32 {
        self.get_value_from_connection(target, connections.get(target).unwrap(), connections)
    }
}

fn solve(connections: &Connections) -> (u32, u32) {
    let part1 = CircuitSolver::new(HashMap::new()).solve_for("a", connections);
    let mut solutions = HashMap::new();
    solutions.insert(String::from("b"), part1);
    (
        part1,
        CircuitSolver::new(solutions).solve_for("a", connections),
    )
}

fn get_input(file_path: &String) -> Connections {
    let source_target_regex = Regex::new(r"^(.*)\s->\s(\w+)$").unwrap();
    let input_regex = Regex::new(r"^[^\s]+$").unwrap();
    let unary_regex = Regex::new(r"NOT\s(\w+)$").unwrap();
    let binary_regex = Regex::new(r"^(\w+|\d+)\s+(AND|OR|LSHIFT|RSHIFT)\s+(\w+|\d+)").unwrap();
    std::fs::read_to_string(file_path)
        .expect("Error reading input file!")
        .lines()
        .map(|line| match source_target_regex.captures(line) {
            Some(cap) => {
                let source = String::from(cap.get(1).unwrap().as_str());
                let target = String::from(cap.get(2).unwrap().as_str());
                if input_regex.is_match(&source) {
                    return (target, Connection::Input(Operand::new(source)));
                }
                if let Some(unary_cap) = unary_regex.captures(&source) {
                    return (
                        target,
                        Connection::Not(Operand::new(String::from(
                            unary_cap.get(1).unwrap().as_str(),
                        ))),
                    );
                }
                if let Some(binary_cap) = binary_regex.captures(&source) {
                    return (
                        target,
                        Connection::Binary(
                            Operand::new(String::from(binary_cap.get(1).unwrap().as_str())),
                            Operand::new(String::from(binary_cap.get(3).unwrap().as_str())),
                            Operation::new(binary_cap.get(2).unwrap().as_str()),
                        ),
                    );
                };
                panic!("Unrecognized operation source: {}", source)
            }
            _ => panic!("Unrecognized operation line: {}", line),
        })
        .collect()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = std::time::Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    let end = now.elapsed().as_secs_f32();
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:7}", end);
}

File: 2015/07/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <regex.h>
#include <string.h>

#define INPUT_INCREMENT 10

typedef enum
{
    SCALAR,
    WIRE
} OperandType;

typedef enum
{
    AND,
    OR,
    LSHIFT,
    RSHIFT
} OperationType;

typedef enum
{
    INPUT,
    NOT,
    BINARY
} ConnectionType;

typedef struct
{
    OperandType type;
    int value;
} Operand;

typedef struct
{
    ConnectionType type;
    Operand operand1, operand2;
    OperationType operation;
} Connection;

typedef struct
{
    int count, capacity;
    char **wires;
    Connection *connections;
} Input;

typedef struct
{
    int part1;
    int part2;
} Results;

typedef struct
{
    int count;
    char **wires;
    Connection *connections;
    int *value;
    int *valuesFound;
} Circuit;

int getWireIndex(Input *input, const char *wire);
int getValueFromConnection(Circuit circuit, int target);

int getValueFromOperand(Circuit circuit, Operand operand)
{
    switch (operand.type)
    {
    case SCALAR:
        return operand.value;
    case WIRE:
        return getValueFromConnection(circuit, operand.value);
    default:
        perror("Unrecognized operand type");
        exit(1);
    }
}

int getValueFromBinaryConnection(Circuit circuit, Connection connection)
{
    int operand1 = getValueFromOperand(circuit, connection.operand1);
    int operand2 = getValueFromOperand(circuit, connection.operand2);
    switch (connection.operation)
    {
    case AND:
        return operand1 & operand2;
    case OR:
        return operand1 | operand2;
    case LSHIFT:
        return operand1 << operand2;
    case RSHIFT:
        return operand1 >> operand2;
    default:
        perror("Unrecognized binary operation");
        exit(1);
    }
}

int calculateValueForConnection(Circuit circuit, Connection connection)
{
    switch (connection.type)
    {
    case INPUT:
        return getValueFromOperand(circuit, connection.operand1);
    case NOT:
        return ~getValueFromOperand(circuit, connection.operand1);
    case BINARY:
        return getValueFromBinaryConnection(circuit, connection);
    default:
        perror("Unrecognized connection type");
        exit(1);
    }
}

int setValueOnCircuit(Circuit circuit, int target, int value)
{
    circuit.valuesFound[target] = 1;
    return circuit.value[target] = value;
}

int getValueFromConnection(Circuit circuit, int target)
{
    if (circuit.valuesFound[target])
        return circuit.value[target];
    return setValueOnCircuit(circuit, target, calculateValueForConnection(circuit, circuit.connections[target]));
}

Circuit createCircuit(Input input)
{
    return (Circuit){
        input.count,
        input.wires,
        input.connections,
        calloc(input.count, sizeof(int)),
        calloc(input.count, sizeof(int))};
}

void freeCircuit(Circuit circuit)
{
    free(circuit.value);
    free(circuit.valuesFound);
}

Results solve(Input input)
{
    Circuit circuit = createCircuit(input);
    int targetWire = getWireIndex(&input, "a");
    int part1Result = getValueFromConnection(circuit, targetWire);
    freeCircuit(circuit);
    circuit = createCircuit(input);
    setValueOnCircuit(circuit, getWireIndex(&input, "b"), part1Result);
    int part2Result = getValueFromConnection(circuit, targetWire);
    freeCircuit(circuit);
    return (Results){part1Result, part2Result};
}

int getWireIndex(Input *input, const char *wire)
{
    int index = 0;
    while (index < input->count)
        if (strcmp(wire, input->wires[index++]) == 0)
            return index - 1;
    if (input->count == input->capacity)
    {
        input->capacity += INPUT_INCREMENT;
        char **oldWires = input->wires;
        char **newWires = realloc(oldWires, input->capacity * sizeof(char *));
        input->wires = newWires;
        Connection *oldConnections = input->connections;
        Connection *newConnections = realloc(oldConnections, input->capacity * sizeof(Connection));
        input->connections = newConnections;
    }
    input->wires[input->count] = malloc(strlen(wire) + 1);
    strcpy(input->wires[input->count], wire);
    input->count++;
    return input->count - 1;
}

void addToInput(Input *input, const char *target, Connection connection)
{
    int wire = getWireIndex(input, target);
    input->connections[wire] = connection;
}

Operand getOperand(Input *input, const char *value)
{
    int scalar;
    if (sscanf(value, "%d", &scalar))
        return (Operand){SCALAR, scalar};
    return (Operand){WIRE, getWireIndex(input, value)};
}

OperationType getOperation(const char *value)
{
    if (strcmp(value, "AND") == 0)
        return AND;
    if (strcmp(value, "OR") == 0)
        return OR;
    if (strcmp(value, "LSHIFT") == 0)
        return LSHIFT;
    if (strcmp(value, "RSHIFT") == 0)
        return RSHIFT;
    perror("Unknow operation");
    perror(value);
    exit(1);
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    regex_t sourceTargetRegex, inputRegex, unaryRegex, binaryRegex;
    if (regcomp(&sourceTargetRegex, "^(.*) -> ([a-z]+)", REG_EXTENDED))
    {
        perror("Error compiling sourceTarget regex.");
        exit(1);
    }
    if (regcomp(&inputRegex, "^[^ ]+$", REG_EXTENDED))
    {
        perror("Error compiling input regex.");
        exit(1);
    }
    if (regcomp(&unaryRegex, "^NOT ([a-z]+)", REG_EXTENDED))
    {
        perror("Error compiling not regex.");
        exit(1);
    }
    if (regcomp(&binaryRegex, "^([a-z]+|[0-9]+) +(AND|OR|LSHIFT|RSHIFT) +([a-z]+|[0-9]+)", REG_EXTENDED))
    {
        perror("Error compiling not regex.");
        exit(1);
    }
    Input input = {
        0, INPUT_INCREMENT,
        malloc(INPUT_INCREMENT * sizeof(char *)),
        malloc(INPUT_INCREMENT * sizeof(Connection))};
#define REGEX_GROUP_COUNT 7
    char *line, *cursor, source[16], target[8];
    size_t lineLength;
    regmatch_t groupArray[REGEX_GROUP_COUNT];
    Operand operand1, operand2;
    OperationType operation;
    int group;
    while (getline(&line, &lineLength, file) != EOF)
    {
        cursor = line;
        if (!regexec(&sourceTargetRegex, cursor, REGEX_GROUP_COUNT, groupArray, 0))
        {
            char cursorCopy[strlen(cursor) + 1];
            strcpy(cursorCopy, cursor);
            for (group = 0; group <= REGEX_GROUP_COUNT && groupArray[group].rm_so != -1; group++)
            {
                cursorCopy[groupArray[group].rm_eo] = 0;
                switch (group)
                {
                case 1:
                    strcpy(source, cursorCopy + groupArray[group].rm_so);
                    break;
                case 2:
                    strcpy(target, cursorCopy + groupArray[group].rm_so);
                    break;
                }
            }
            cursor = source;
            strcpy(cursorCopy, cursor);
            if (!regexec(&inputRegex, cursor, REGEX_GROUP_COUNT, groupArray, 0))
                addToInput(&input, target, (Connection){INPUT, getOperand(&input, source)});
            else if (!regexec(&unaryRegex, cursor, REGEX_GROUP_COUNT, groupArray, 0))
            {
                for (group = 0; group <= REGEX_GROUP_COUNT && groupArray[group].rm_so != -1; group++)
                {
                    cursorCopy[groupArray[group].rm_eo] = 0;
                    switch (group)
                    {
                    case 1:
                        operand1 = getOperand(&input, cursorCopy + groupArray[group].rm_so);
                        break;
                    }
                }
                addToInput(&input, target, (Connection){NOT, operand1});
            }
            else if (!regexec(&binaryRegex, cursor, REGEX_GROUP_COUNT, groupArray, 0))
            {
                for (group = 0; group <= REGEX_GROUP_COUNT && groupArray[group].rm_so != -1; group++)
                {
                    cursorCopy[groupArray[group].rm_eo] = 0;
                    switch (group)
                    {
                    case 1:
                        operand1 = getOperand(&input, cursorCopy + groupArray[group].rm_so);
                        break;
                    case 2:
                        operation = getOperation(cursorCopy + groupArray[group].rm_so);
                        break;
                    case 3:
                        operand2 = getOperand(&input, cursorCopy + groupArray[group].rm_so);
                        break;
                    }
                }
                addToInput(
                    &input,
                    target,
                    (Connection){BINARY, operand1, operand2, operation});
            }
            else
            {
                perror("Unknow connection source");
                perror(source);
                exit(1);
            }
        }
        else
        {
            perror("Bad format line");
            perror(line);
            exit(1);
        }
    }
    fclose(file);
    return input;
}

void freeInput(Input input)
{
    int index = 0;
    while (index < input.count)
        free(input.wires[index++]);
    free(input.wires);
    free(input.connections);
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}